### PR TITLE
[WebRTC][WPE][GTK] SFrame encryption support

### PR DESCRIPTION
**Pull request by philn**
#### 8cf3d3cdb4d73ecfd90b0e44d60fec1053c2df48
<pre>
[WebRTC][WPE][GTK] SFrame encryption support
<a href="https://bugs.webkit.org/show_bug.cgi?id=277602">https://bugs.webkit.org/show_bug.cgi?id=277602</a>

Reviewed by Michael Catanzaro.

OpenSSL is now used for SFrame crypto handling in the GTK and WPE ports. GStreamer encoded
transforms handling will be submitted in a future PR.

* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformer.cpp:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformerCocoa.cpp:
(WebCore::encodeBigEndian): Deleted.
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformerOpenSSL.cpp: Added.
(WebCore::deriveHKDFBits):
(WebCore::RTCRtpSFrameTransformer::computeSaltKey):
(WebCore::createBaseSFrameKey):
(WebCore::RTCRtpSFrameTransformer::computeAuthenticationKey):
(WebCore::RTCRtpSFrameTransformer::computeEncryptionKey):
(WebCore::crypt):
(WebCore::RTCRtpSFrameTransformer::decryptData):
(WebCore::RTCRtpSFrameTransformer::encryptData):
(WebCore::RTCRtpSFrameTransformer::computeEncryptedDataSignature):
(WebCore::RTCRtpSFrameTransformer::updateAuthenticationSize):
* Source/WebCore/Modules/mediastream/SFrameUtils.h:
(WebCore::encodeBigEndian):
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:

Canonical link: <a href="https://commits.webkit.org/282124@main">https://commits.webkit.org/282124@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/46fee37f9aca2c5571daa906b84ee9576b732781

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62168 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41522 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14760 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66148 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12713 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64287 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49209 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13053 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50099 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/8797 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65237 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38567 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/53845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30901 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35245 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/11113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11644 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/55264 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56975 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/11417 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67878 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/61410 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6111 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/57475 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6137 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/53823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57699 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13813 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5048 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83174 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/37322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14584 "Found 1 new JSC stress test failure: wasm.yaml/wasm/fuzz/memory.js.wasm-collect-continuously (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38406 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39502 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/38151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

**Comment by philn** [comment]
@mcatanzaro are you happier? :)

**Comment by mcatanzaro** [comment]
Heh, I had contacted you on Matrix to point you to https://bugs.webkit.org/show_bug.cgi?id=216201 but I didn't expect you to actually rewrite the pull request with that in mind. Cool. Thanks.

But it is slightly weird that this works, because we don't explicitly link to OpenSSL anywhere, do we? It's surely only working by coincidence, because gst-webrtc depends on OpenSSL?

**Comment by philn** [comment]
Indeed, good point. I'll update the PR.

**Comment by philn** [comment]
@mcatanzaro r?

**Comment by philn** [comment]
> I can give you a rubber stamp review. You sure you don't want a multimedia reviewer for this?

This part isn't much related with multimedia. The next one will though, but that's for later :)

File: Source/cmake/GStreamerChecks.cmake [hunk 1]
```diff
@@ -60,14 +60,15 @@ endif ()
...
-            message(FATAL_ERROR "OpenSSL 3 is needed for USE_GSTREAMER_WEBRTC.")
-        endif ()
     else ()
         SET_AND_EXPOSE_TO_BUILD(USE_LIBWEBRTC TRUE)
     endif ()
+
+    # OpenSSL is required for SFrame encryption support.
+    find_package(OpenSSL)
+    if (NOT OPENSSL_FOUND OR OPENSSL_VERSION VERSION_LESS "3.0.0")
+        message(FATAL_ERROR "OpenSSL 3 is needed for SFrame encryption support.")
```
Comment (mcatanzaro) [inline Source/cmake/GStreamerChecks.cmake]: What's important here is to say what build option to turn off if you don't have OpenSSL 3 available, not to say what precisely it's being used for:

```suggestion
        message(FATAL_ERROR "OpenSSL 3 is needed for ENABLE_WEB_RTC.")
```

**Review by mcatanzaro — APPROVED** [review]
I'm not familiar with OpenSSL, but I don't notice any problems.

**Pull request merged**